design/0001-metallb-crd.md

```diff
@@ -0,0 +1,251 @@
+# MetalLB Custom Resource Controller
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Initial CRD definition](#initial-crd-definition)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+
+<!-- /toc -->
+
+## Summary
+
+The purpose of this enhancement is to be able to use kubernetes Custom
+Resource Definition "CRD" to configure MetalLB layer2 and BGP.
+
+## Motivation
+
+Today MetalLB allows only configuration of layer2 and BGP protocols using
+ConfigMap objects, which is simple from configuration perspective and doesn't
+require cluster-admin rights, however it doesn't have validation on the API
+server level, no support for API versioning and no fine-grained permission
+model via RBAC.
+Custom resource definition "CRD" takes advantage of kubernetes extensibility
+to deliver the automation advantages of cloud services like provisioning,
+scaling and backup/restore.
+
+### Goals
+
+Configuring MetalLB layer2 and BGP using kubernetes custom resource "CRD"
```

Comment:
I think throughout this enhancement you can just say "configuring MetalLB" and not include "layer2 and BGP".  I think it can be assumed that all MetalLB configuration should be supported through this interface.